packages/babel-helpers/src/helpers/usingCtx.ts

```diff
@@ -104,21 +110,40 @@ export default function _usingCtx(): UsingCtxReturn {
     >(null, true),
     // dispose
     d: function () {
-      var error = this.e;
+      var error = this.e,
+        state: StateFlag = StateFlag.NONE,
+        resource;
 
       function next(): Promise<void> | void {
-        // eslint-disable-next-line @typescript-eslint/no-use-before-define
         while ((resource = stack.pop())) {
           try {
-            var resource,
-              disposalResult = resource.d && resource.d.call(resource.v);
-            if (resource.a) {
-              return Promise.resolve(disposalResult).then(next, err);
+            if (!resource.a && state === StateFlag.NEEDS_AWAIT) {
+              state = StateFlag.NONE;
```

Comment:
Note: If https://github.com/tc39/proposal-explicit-resource-management/pull/219#issuecomment-2177060304 gets consensus, we can simply change this line to `state |= StateFlag.HAS_AWAITED`. Before that happens we implement the 2024-06 version as-is.